crates/miden-objects/src/block/proposed_block.rs

```diff
@@ -0,0 +1,563 @@
+                reference_block_num: batch.reference_block_num(),
+                batch_id: batch.id(),
+            });
+        }
+    }
+
+    Ok(())
+}
+
+/// Computes the [`BlockNoteTree`] from the note trees of the batches in the block.
+///
+/// This takes the batch note tree of a batch and removes any note that was erased (i.e. consumed by
+/// another batch in the block), as dictated by the `block_output_notes` map.
+///
+/// Then inserts the batch note tree as a subtree into the larger block note tree.
+fn compute_block_note_tree(
+    batches: &[ProvenBatch],
+    block_output_notes: &BTreeMap<NoteId, (BatchId, OutputNote)>,
+) -> BlockNoteTree {
+    let mut block_note_tree = BlockNoteTree::empty();
+
+    for (batch_idx, batch) in batches.iter().enumerate() {
+        let mut batch_output_notes_tree = batch.output_notes_tree().clone();
+
+        for (note_tree_idx, original_output_note) in batch.output_notes().iter().enumerate() {
+            // Note that because we disallow duplicate output notes, if this map contains a note id,
+            // then we can be certain it was created by this batch and should stay in the tree.
+            if !block_output_notes.contains_key(&original_output_note.id()) {
+                let index = u64::try_from(note_tree_idx).expect(
+                  "the number of output notes should be less than MAX_OUTPUT_NOTES_PER_BATCH and thus fit into a u64",
+              );
+                batch_output_notes_tree
+                    .remove(index)
+                    .expect("the index should be less than MAX_OUTPUT_NOTES_PER_BATCH");
+            }
+        }
+
+        let batch_idx = u64::try_from(batch_idx)
+            .expect("the index should be less than MAX_BATCHES_PER_BLOCK and thus fit into a u64");
+        block_note_tree
+            .insert_batch_note_subtree(batch_idx, batch_output_notes_tree)
+            .expect("the batch note tree depth should not exceed the block note tree depth and the index should be less than MAX_BATCHES_PER_BLOCK");
+    }
+
+    block_note_tree
+}
+
+// ACCOUNT UPDATE AGGREGATOR
+// ================================================================================================
+
+/// Aggregate all updates for the same account and store each update indexed by its initial
+/// state commitment so we can easily retrieve them later.
+/// This lets us chronologically order the updates per account across batches.
+fn aggregate_account_updates(
+    account_updates: BTreeMap<AccountId, AccountWitness>,
+    batches: &[ProvenBatch],
+) -> Result<UpdatedAccounts, ProposedBlockError> {
+    let mut update_aggregator = AccountUpdateAggregator::new();
+
+    for batch in batches {
+        for (account_id, update) in batch.account_updates() {
+            update_aggregator.insert_update(*account_id, batch.id(), update.clone())?;
+        }
+    }
+
+    update_aggregator.aggregate_all(account_updates)
+}
```

Comment:
nit: maybe instead of this method we could have `AccountUpdateAggregator::from_batches()` which would basically do the stuff in the loop. Then, instead of calling this method we'd do something like:
```rust
let aggregator = AccountUpdateAggregator::from_batches(&batches)?;
let updates = aggregator.into_updates(account_witnesses)?;
```